core/shared/src/main/scala/zio/Semaphore.scala

```diff
@@ -18,153 +18,173 @@ package zio
+
+    private def reserve(n: Long)(implicit trace: Trace): UIO[Reservation] = {
+      if (n <= 0) ZIO.succeedNow(Reservation(ZIO.unit, ZIO.unit))
+      else Promise.make[Nothing, Unit].flatMap { promise =>
+        ZIO.succeed {
+          var loop = true
+          var result: Reservation = null
+          while (loop) {
+            val current = state.get()
+            current match {
+              case Right(permits) if permits >= n =>
+                if (state.compareAndSet(current, Right(permits - n))) {
+                  result = Reservation(ZIO.unit, ZIO.succeed(releaseInternal(n)))
+                  loop = false
+                }
+              case Right(permits) =>
+                val newQueue = ScalaQueue(promise -> (n - permits))
+                if (state.compareAndSet(current, Left(newQueue))) {
+                  result = Reservation(promise.await, ZIO.succeed(releaseInternal(n)))
+                  loop = false
                 }
-            case Right(permits) => ZIO.unit -> Right(permits + n)
-          }.flatten
-
-        def releaseN(n: Long)(implicit trace: Trace): UIO[Any] = {
-
-          @tailrec
-          def loop(
-            n: Long,
-            state: Either[ScalaQueue[(Promise[Nothing, Unit], Long)], Long],
-            acc: UIO[Any]
-          ): (UIO[Any], Either[ScalaQueue[(Promise[Nothing, Unit], Long)], Long]) =
-            state match {
-              case Right(permits) => acc -> Right(permits + n)
               case Left(queue) =>
-                queue.dequeueOption match {
-                  case None => acc -> Right(n)
-                  case Some(((promise, permits), queue)) =>
-                    if (n > permits)
-                      loop(n - permits, Left(queue), acc *> promise.succeedUnit)
-                    else if (n == permits)
-                      (acc *> promise.succeedUnit) -> Left(queue)
-                    else
-                      acc -> Left((promise -> (permits - n)) +: queue)
+                val newQueue = queue.enqueue(promise -> n)
+                if (state.compareAndSet(current, Left(newQueue))) {
+                  result = Reservation(promise.await, ZIO.succeed(releaseInternal(n)))
+                  loop = false
                 }
             }
+          }
+          result
+        }
+      }
+    }
+
+    @tailrec
+    private def releaseInternal(n: Long): Unit = {
+      val current = state.get()
+      current match {
+        case Right(permits) =>
+          if (!state.compareAndSet(current, Right(permits + n))) releaseInternal(n)
+        case Left(queue) =>
+          queue.dequeueOption match {
+            case None =>
+              if (!state.compareAndSet(current, Right(n))) releaseInternal(n)
+            case Some(((promise, required), remaining)) =>
+              if (n >= required) {
+                promise.succeedUnit.unsafe.run
+                if (!state.compareAndSet(current, Left(remaining))) releaseInternal(n - required)
```

Comment:
I don't think this is right. 

If `n > required`, and the `compareAndSet` succeeds, then we still need to release `n -  required` permits.

If `compareAndSet` succeeded, then either another thread _also_ released enough permits and fulfilled the promise, or another thread reduced the number of required permits for the current promise, and we need to take that into account. In either case I _think_ that we can just call `releaseInternal(n)` again. That may end up calling `succeedUnit` on the promise again, but that should be idempotent.

In short, I think this should be:
```suggestion
                if (state.compareAndSet(current, Left(remaining))) {
                  releaseInternal(n - required)
                } else {
                  releaseInternal(n)
                }
```

But I'm not 100% sure there isn't a race condition there that I haven't thought of.